src/rgw/rgw_crypt.cc

```diff
@@ -0,0 +1,1314 @@
+// -*- mode:C; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/**
+ * Crypto filters for Put/Post/Get operations.
+ */
+#include <rgw/rgw_op.h>
+#include <rgw/rgw_crypt.h>
+#include <auth/Crypto.h>
+#include <rgw/rgw_b64.h>
+#include <rgw/rgw_rest_s3.h>
+#include "include/assert.h"
+#include <boost/utility/string_ref.hpp>
+#include <rgw/rgw_keystone.h>
+#include "include/str_map.h"
+#include "../isa-l_crypto_plugin/crypto_accel.h"
+#include "../isa-l_crypto_plugin/crypto_plugin.h"
+
+#ifdef USE_NSS
+# include <nspr.h>
+# include <nss.h>
+# include <pk11pub.h>
+#endif
+
+#ifdef USE_CRYPTOPP
+#include <crypto++/cryptlib.h>
+#include <crypto++/modes.h>
+#include <crypto++/aes.h>
+using namespace CryptoPP;
+#endif
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rgw
+
+using namespace rgw;
+
+/**
+ * Encryption in CTR mode. offset is used as IV for each block.
+ */
+class AES_256_CTR : public BlockCrypt {
+public:
+  static const size_t AES_256_KEYSIZE = 256 / 8;
+  static const size_t AES_256_IVSIZE = 128 / 8;
+private:
+  static const uint8_t IV[AES_256_IVSIZE];
+  CephContext* cct;
+  uint8_t key[AES_256_KEYSIZE];
+public:
+  AES_256_CTR(CephContext* cct): cct(cct) {
+  }
+  ~AES_256_CTR() {
+  }
+  bool set_key(const uint8_t* _key, size_t key_size) {
+    if (key_size != AES_256_KEYSIZE) {
+      return false;
+    }
+    memcpy(key, _key, AES_256_KEYSIZE);
+    return true;
+  }
+  size_t get_block_size() {
+    return AES_256_IVSIZE;
+  }
+
+#ifdef USE_CRYPTOPP
+
+  bool encrypt(bufferlist& input, off_t in_ofs, size_t size, bufferlist& output, off_t stream_offset) {
+    byte iv[AES_256_IVSIZE];
+    ldout(cct, 25)
+        << "Encrypt in_ofs " << in_ofs
+        << " size=" << size
+        << " stream_offset=" << stream_offset
+        << dendl;
+    if (input.length() < in_ofs + size) {
+      return false;
+    }
+
+    output.clear();
+    buffer::ptr buf((size + AES_256_KEYSIZE - 1) / AES_256_KEYSIZE * AES_256_KEYSIZE);
+    /*create CTR mask*/
+    prepare_iv(iv, stream_offset);
+    CTR_Mode< AES >::Encryption e;
+    e.SetKeyWithIV(key, AES_256_KEYSIZE, iv, AES_256_IVSIZE);
+    buf.zero();
+    e.ProcessData((byte*)buf.c_str(), (byte*)buf.c_str(), buf.length());
+    buf.set_length(size);
+    off_t plaintext_pos = in_ofs;
+    off_t crypt_pos = 0;
+    auto iter = input.buffers().begin();
+    //skip unaffected begin
+    while ((iter != input.buffers().end()) && (plaintext_pos >= iter->length())) {
+      plaintext_pos -= iter->length();
+      ++iter;
+    }
+    while (iter != input.buffers().end()) {
+      off_t cnt = std::min((off_t)(iter->length() - plaintext_pos), (off_t)(size - crypt_pos));
```

Comment:
you can call `std::min<off_t>()` explicitly to avoid duplicating the cast to (off_t):
```c++
off_t cnt = std::min<off_t>(iter->length() - plaintext_pos, size - crypt_pos);
```